homeassistant/components/matrix.py

```diff
@@ -0,0 +1,259 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+SESSION_FILE = 'matrix.conf'
+
+CONF_HOMESERVER = 'homeserver'
+CONF_LISTENING_ROOMS = 'listening_rooms'
+CONF_COMMANDS = 'commands'
+
+EVENT_MATRIX_COMMAND = 'matrix_command'
+
+DOMAIN = 'matrix'
+
+PLATFORM_SCHEMA = vol.Schema({
+    vol.Required(CONF_HOMESERVER): cv.url,
+    vol.Optional(CONF_VERIFY_SSL, default=True): cv.boolean,
+    vol.Required(CONF_USERNAME): cv.string,
+    vol.Required(CONF_PASSWORD): cv.string,
+    vol.Optional(CONF_LISTENING_ROOMS, default=[]): vol.All(cv.ensure_list,
+                                                            [cv.string]),
+    vol.Optional(CONF_COMMANDS, default=[]): vol.All(cv.ensure_list,
+                                                     [cv.string]),
+})
+
+SERVICE_SEND_MESSAGE = 'send_message'
+
+SERVICE_SCHEMA_SEND_MESSAGE = vol.Schema({
+    vol.Required(ATTR_MESSAGE): cv.string,
+    vol.Required(ATTR_TARGET): vol.All(cv.ensure_list, [cv.string]),
+})
+
+
+def setup(hass, config):
+    """Set up the Matrix bot component."""
+    if not config[DOMAIN]:
+        return False
+
+    config = config[DOMAIN][0]
+
+    bot = MatrixBot(
+        hass,
+        os.path.join(hass.config.path(), SESSION_FILE),
+        config.get(CONF_HOMESERVER),
+        config.get(CONF_VERIFY_SSL),
+        config.get(CONF_USERNAME),
+        config.get(CONF_PASSWORD),
+        config.get(CONF_LISTENING_ROOMS, []),
+        config.get(CONF_COMMANDS, []))
+    hass.data[DOMAIN] = bot
+
+    def send_message_handler(service):
+        """Handle the send_message service."""
+        bot.handle_send_message(service)
+
+    hass.services.async_register(
+        DOMAIN, SERVICE_SEND_MESSAGE, send_message_handler,
+        schema=SERVICE_SCHEMA_SEND_MESSAGE)
+
+    return True
+
+
+class MatrixBot(object):
+    """The Matrix Bot."""
+
+    def __init__(self, hass, config_file, homeserver, verify_ssl,
+                 username, password, listening_rooms, commands):
+        """Set up the client."""
+        self.hass = hass
+
+        self._session_filepath = config_file
+        self._auth_tokens = self._get_auth_tokens()
+
+        self._homeserver = homeserver
+        self._verify_tls = verify_ssl
+        self._username = username
+        self._password = password
+
+        self._mx_id = "{user}@{homeserver}".format(
+            user=username, homeserver=urlparse(homeserver).netloc)
+
+        self._listening_rooms = listening_rooms
+        self._commands = set(commands)
+
+        # Login, this will raise a MatrixRequestError if login is unsuccessful
+        self._client = self._login()
+
+        # Join rooms in which we listen for commands and start listening
+        self._join_rooms()
+        self._client.start_listener_thread()
+
+    def _join_rooms(self):
+        """Join the rooms that we listen for commands in."""
+        from matrix_client.client import MatrixRequestError
+
+        def room_message_cb(room, event):
```

Comment:
Could this function not be a method on the main class rather than nested in here?

Reply:
Yes, maybe that's nicer. I'll change that!